test/stdlib_dependencies.jl

```diff
@@ -0,0 +1,197 @@
+using Pkg, Test, Libdl
+
+# Remove `.X.dylib` or just `.dylib`
+function strip_soversion_macos(lib)
+    m = match(r"^(.*?)(\.\d+)*\.dylib$", lib)
+    if m !== nothing
+        return m.captures[1]
+    end
+    return lib
+end
+
+# Remove `.so.X` or just `.so`
+function strip_soversion_linux(lib)
+    m = match(r"^(.*?)\.so(\.\d+)*$", lib)
+    if m !== nothing
+        return m.captures[1]
+    end
+    return lib
+end
+
+# Remove `-X.dll` or just `.dll`
+function strip_soversion_windows(lib)
+    m = match(r"^(.*?)(-\d+)*\.dll$", lib)
+    if m !== nothing
+        return m.captures[1]
+    end
+    return lib
+end
+
+
+function get_deps_otool(lib_path::String)
+    libs = split(readchomp(`otool -L $(lib_path)`), "\n")[2:end]
+    # Get rid of `(compatibility version x.y.x)` at the end
+    libs = first.(split.(libs, (" (compatibility",)))
+    # Get rid of any `@rpath/` stuff at the beginning
+    libs = last.(split.(libs, ("@rpath/",)))
+
+    # If there are any absolute paths left, get rid of them here
+    libs = basename.(strip.(libs))
+
+    # Now that we've got the basenames of each library, remove `.x.dylib` if it exists:
+    libs = strip_soversion_macos.(libs)
+
+    # Get rid of any self-referential links
+    self_lib = strip_soversion_macos(basename(lib_path))
+    libs = filter(l -> l != self_lib, libs)
+    return libs
+end
+
+function is_system_lib_macos(lib)
+    system_libs = [
+        "libSystem.B",
+        "libc++", # While we package libstdc++, we do NOT package libc++.
+        "libiconv", # some things (like git) link against system libiconv
+
+        # macOS frameworks used by things like LibCurl
+        "CoreFoundation",
+        "CoreServices",
+        "Security",
+        "SystemConfiguration"
+    ]
+    return lib ∈ system_libs
+end
+
+function is_system_lib_linux(lib)
+    system_libs = [
+        "libdl",
+        "libc",
+        "libm",
+        "librt",
+        "libpthread",
+        "ld-linux-x86-64",
+        "ld-linux-x86",
+        "ld-linux-aarch64",
+        "ld-linux-armhf",
+        "ld-linux-i386",
+    ]
+    return lib ∈ system_libs
+end
+
+function is_system_lib_freebsd(lib)
+    system_libs = [
+        "libc",
+        "libm",
+        "libthr",      # primary threading library
+        "libpthread",  # alias kept for compatibility
+        "librt",
+        "libutil",
+        "libexecinfo",
+        "libcxxrt",
+    ]
+    return lib ∈ system_libs
+end
+
+function get_deps_readelf(lib_path::String)
+    # Split into lines
+    libs = split(readchomp(`readelf -d $(lib_path)`), "\n")
+
+    # Only keep `(NEEDED)` lines
+    needed_str = Sys.isfreebsd() ? "NEEDED" : "(NEEDED)"
+    libs = filter(occursin(needed_str), libs)
```

Comment:
That's not what I suggested: https://github.com/JuliaLang/julia/pull/58444#discussion_r2105152619 :wink: